### PR TITLE
Allow reduction of HVDC lines

### PR DESCRIPTION
**Pull request by HugoKulesza**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

Issue #2401

### CODE REVIEW

File: iidm/iidm-reducer/src/main/java/com/powsybl/iidm/reducer/DefaultNetworkReducer.java [hunk 1]
```diff
@@ -178,6 +182,57 @@ private Load replaceBranchByLoad(Branch<?> branch, VoltageLevel vl, Terminal ter
         return load;
     }
 
+    private void replaceHvdcLine(HvdcLine hvdcLine, VoltageLevel vl, Terminal terminal, HvdcConverterStation.HvdcType type) {
+        if (type == HvdcConverterStation.HvdcType.VSC) {
```
Comment (geofjamg) [inline iidm/iidm-reducer/src/main/java/com/powsybl/iidm/reducer/DefaultNetworkReducer.java:186]: This this the invert and only if the VSC has the boolean voltageRegulatorOn to true. LCC are always replaced by a load.

Comment (HugoKulesza) [reply iidm/iidm-reducer/src/main/java/com/powsybl/iidm/reducer/DefaultNetworkReducer.java]: Changes are done (if I understood right).

**Comment by geofjamg** [comment]
> Note: the DC converter station is not removed when the HVDC line is reduced, shouldn't it be? @geofjamg @sylvlecl

Yes it should.

**Review by miovd — APPROVED** [review]

File: iidm/iidm-reducer/src/main/java/com/powsybl/iidm/reducer/DefaultNetworkReducer.java [hunk 2]
```diff
@@ -96,20 +96,24 @@ protected void reduce(ThreeWindingsTransformer transformer) {
...
-        VoltageLevel vl2 = hvdcLine.getConverterStation2().getTerminal().getVoltageLevel();
-        if (getPredicate().test(vl1) || getPredicate().test(vl2)) {
-            throw new UnsupportedOperationException("Reduction of HVDC lines is not supported");
+        Terminal terminal1 = hvdcLine.getConverterStation1().getTerminal();
+        Terminal terminal2 = hvdcLine.getConverterStation2().getTerminal();
+        VoltageLevel vl1 = terminal1.getVoltageLevel();
+        VoltageLevel vl2 = terminal2.getVoltageLevel();
+        HvdcConverterStation<?> station1 = hvdcLine.getConverterStation1();
+        HvdcConverterStation<?> station2 = hvdcLine.getConverterStation2();
+        if (getPredicate().test(vl1)) {
```
Comment (geofjamg) [inline iidm/iidm-reducer/src/main/java/com/powsybl/iidm/reducer/DefaultNetworkReducer.java:110]: We should test first if vl1 AND vl1 have to be removed ?

Comment (geofjamg) [reply iidm/iidm-reducer/src/main/java/com/powsybl/iidm/reducer/DefaultNetworkReducer.java:110]: My question is just: does it work if both voltage levels have to be removed?

Comment (HugoKulesza) [reply iidm/iidm-reducer/src/main/java/com/powsybl/iidm/reducer/DefaultNetworkReducer.java:110]: If both voltage levels have to be removed, test(vl1) = test(vl2) = false so we remove the hvdcLine (l.115).
However I have to add here the removal of converter stations as well.

Comment (geofjamg) [reply iidm/iidm-reducer/src/main/java/com/powsybl/iidm/reducer/DefaultNetworkReducer.java:110]: maybe in that case, it is automatically removed with the voltage level?

Comment (HugoKulesza) [reply iidm/iidm-reducer/src/main/java/com/powsybl/iidm/reducer/DefaultNetworkReducer.java:110]: In fact the last comment is not necessary : converter stations are in this case already removed when removing the voltage levels.

**Review by geofjamg — APPROVED** [review]

**Pull request merged**